internal/operator/internal/operator_installer.go

```diff
@@ -62,3 +66,67 @@ func (o OperatorInstaller) InstallOperator(ctx context.Context) (*v1alpha1.Clust
 
 	return todo, nil
 }
+
+func (o OperatorInstaller) verifyOperatorInstallation(ctx context.Context, ip *v1alpha1.InstallPlan) (*v1alpha1.ClusterServiceVersion, error) {
+	ipKey := types.NamespacedName{
+		Namespace: ip.GetNamespace(),
+		Name:      ip.GetName(),
+	}
+	if err := wait.PollImmediateUntil(time.Millisecond*250, func() (bool, error) {
+		if err := o.cfg.Client.Get(ctx, ipKey, ip); err != nil {
+			return false, err
+		}
+		if ip.Status.Phase == v1alpha1.InstallPlanPhaseComplete {
+			return true, nil
+		}
+		return false, nil
+	}, ctx.Done()); err != nil {
+		return nil, fmt.Errorf("waiting for operator installation to complete: %v", err)
+	}
+
+	// verify CSV
+	csv, err := o.verifyCSV(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("waiting for CSV installation to complete: %v", err)
+	}
+	return csv, nil
+}
+
+func (o OperatorInstaller) verifyCSV(ctx context.Context) (*v1alpha1.ClusterServiceVersion, error) {
+
+	csvKey := types.NamespacedName{
+		Namespace: o.cfg.Namespace,
+	}
+
+	if csvKey.Name == "" {
+		return nil, fmt.Errorf("could not find installed CSV in install plan")
+	}
+
+	csv := &v1alpha1.ClusterServiceVersion{}
+	if err := wait.PollImmediateUntil(time.Millisecond*250, func() (bool, error) {
+		if err := o.cfg.Client.Get(ctx, csvKey, csv); err != nil {
+			return false, fmt.Errorf("get clusterserviceversion: %v", err)
+		}
+		if csv.Status.Phase == v1alpha1.CSVPhaseSucceeded {
+			return true, nil
+		}
+		return false, nil
+	}, ctx.Done()); err != nil {
+		// Check deplpoyment status
+		dep := &appsv1.Deployment{}
+		depName := csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs[0].Name
+		depKey := types.NamespacedName{
+			Namespace: csvKey.Namespace,
+			Name:      depName,
+		}
+		if err := o.cfg.Client.Get(ctx, depKey, dep); err != nil {
+			return nil, fmt.Errorf("get deployment: %v", err)
+		}
+		for _, s := range dep.Status.Conditions {
+			if s.Type != "Progressing" && s.Type != "Available" {
+				return nil, fmt.Errorf("Deployment failed: %v", s.Message)
+			}
+		}
```

Comment:
I ran a quick test by creating a deployment with an image I know doesn't exist. The deployment conditions I see are:
```yaml
  conditions:
  - lastTransitionTime: "2020-08-26T17:11:29Z"
    lastUpdateTime: "2020-08-26T17:11:29Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
  - lastTransitionTime: "2020-08-26T17:11:29Z"
    lastUpdateTime: "2020-08-26T17:11:29Z"
    message: ReplicaSet "test-dep-97bfdcd79" is progressing.
    reason: ReplicaSetUpdated
    status: "True"
    type: Progressing
```

And the underlying pod has this:
```yaml
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2020-08-26T17:11:29Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2020-08-26T17:11:29Z"
    message: 'containers with unready status: [test-fake-jwl-image]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2020-08-26T17:11:29Z"
    message: 'containers with unready status: [test-fake-jwl-image]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2020-08-26T17:11:29Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: test-fake-jwl-image
    imageID: ""
    lastState: {}
    name: test-fake-jwl-image
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        message: Back-off pulling image "test-fake-jlanford-image"
        reason: ImagePullBackOff
```

We probably need to be checking [all of the deployment conditions](https://github.com/kubernetes/api/blob/051d027c14e16c7a01fab9b432cfad59618000cf/apps/v1/types.go#L418-L430) and looking for any of them that are indicative of failure. And then if we find one, perhaps we could try digging further, such that we eventually end up being able to tell the user something like:

`failed to run operator: timed out waiting for csv success: deployment unavailable: pod is Pending with ImagePullBackOff: Back-off pulling image "test-fake-jlanford-image"`

Reply:
So, DeploymentConstions of "Progressing" and "Available" are not considered failure. In current code , we are considering as failure, if the condition is anything else 
```go
		for _, s := range dep.Status.Conditions {
			if s.Type != "Progressing" && s.Type != "Available" {
				return nil, fmt.Errorf("Deployment failed: %v", s.Message)
			}
```
Currently, I am returning the `Message` filed in such case. However, as you have suggested to dig deeper, we need to dig into pods, and grab the error message from there. Is that what you meant?

Reply:
>  So, DeploymentConditions of Progressing and Available are not considered failure

We have to look at the condition status too. For example, Available = False and Progressing = False mean something's not working correctly.